freestyle-cache/shared/src/main/scala/cache.scala

```diff
@@ -0,0 +1,117 @@
+package freestyle
+
+import cats.{~>}
+
+package cache {
+
+  class KeyValueProvider[Key, Val] {
+
+    /* CacheM trait is a type-class of functors for which key-value store operations
+     * can be provided.
+     *
+     * We assume that the actual store is too big or remote to allow for general
+     * operations over all values, or to search or to iterate over all keys
+     */
+    @free sealed trait CacheM[F[_]] {
+
+      // Gets the value associated to a key, if there is one */
+      def get(key: Key): FreeS.Par[F, Option[Val]]
+
+      // Sets the value of a key to a newValue.
+      def put(key: Key, newVal: Val): FreeS.Par[F, Unit]
```

Comment:
The methods `putAll` and `putIfAbsent` are straight-forward. AAMOF, the latter is directly supported in [redis](https://redis.io/commands/setnx). The command `isEmpty` can be supported efficiently through the use of the [`scan` redis command](https://redis.io/commands/scan). The `containsValue` is the one that could be less efficient to implement.